types.go

```diff
@@ -0,0 +1,449 @@
+// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package hujson contains a parser and packer for the HuJSON format.
+//
+// HuJSON is an extension of standard JSON (as defined in RFC 8259) in order to
+// make it more suitable for humans and configuration files. In particular,
+// it supports line comments (e.g., //...), block comments (e.g., /*...*/), and
+// trailing commas after the last member or element in a JSON object or array.
+//
+//
+// Functionality
+//
+// The Parse function parses HuJSON input as a Value,
+// which is a syntax tree exactly representing the input.
+// Comments and whitespace are represented using the Extra type.
+// Composite types in JSON are represented using the Object and Array types.
+// Primitive types in JSON are represented using the Literal type.
+// The Value.Pack method serializes the syntax tree as raw output,
+// which is byte-for-byte identical to the input if no transformations
+// were performed on the value.
+//
+// A HuJSON value can be transformed using the Minimize, Standardize,
+// NormalizeNames, and Reformat methods. Each of these methods mutate the value
+// in place. Call the Clone method in order to preserve the original value.
+// The Minimize and Standardize methods coerces HuJSON into valid standard JSON.
+// The Reformat method formats the value; it is similar to `go fmt`,
+// but instead for the HuJSON and standard JSON format.
+//
+//
+// Grammar
+//
+// The changes to the JSON grammar are:
+//
+//	--- grammar.json
+//	+++ grammar.hujson
+//	@@ -1,13 +1,31 @@
+//	 members
+//	 	member
+//	+	member ',' ws
+//	 	member ',' members
+//
+//	 elements
+//	 	element
+//	+	element ',' ws
+//	 	element ',' elements
+//
+//	+comments
+//	+	"*/"
+//	+	comment comments
+//	+
+//	+comment
+//	+	'0000' . '10FFFF'
+//	+
+//	+linecomments
+//	+	'\n'
+//	+	linecomment linecomments
+//	+
+//	+linecomment
+//	+	'0000' . '10FFFF' - '\n'
+//	+
+//	 ws
+//	 	""
+//	+	"/*" comments
+//	+	"//" linecomments
+//	 	'0020' ws
+//	 	'000A' ws
+//	 	'000D' ws
+//
+package hujson
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"math"
+	"strconv"
+	"unicode/utf8"
+)
+
+// Value is an exact syntactic representation of a JSON value.
+// The starting and ending byte offsets are populated when parsing,
+// but are otherwise ignored when packing.
+type Value struct {
```

Comment:
I worry about this being used as a value type everywhere in the API. It has exported pointer-containing types (the Extra slice) so there can be aliasing problems.

https://github.com/golang/go/wiki/CodeReviewComments#receiver-type would say the API should work in terms of `*Value`. Or have things unexported and immutable.

Reply:
If we go with `*Value`, then the natural representation for `Object` and `Array` would be to represent the values as a `[]*Value`. This means allocating many small `Value` objects and also means that we have to deal with what it means for a value to be nil.

In the code, I use `*Value` if I need to mutate the value, and `Value` otherwise. The latter is not a guarantee of being read-only, but a helpful signal.

Reply:
If you want to keep it as-is, at least document the risk and maybe add a Clone method on the value receiver type of Value to return a Value value type that doesn't alias?

Reply:
I converted everything to use `*Value` everything and personally felt it was more cumbersome in a number of cases. I'll document the risks.

> add a Clone method on the value receiver type of Value to return a Value value type that doesn't alias?

Already exists. I mentioned it in the documentation